File: code-awesome/python-flask-react/super_dating/backend/src/pages/comment.py
# prg3/python-flask-react/super_dating/backend/src/pages/comment.py
"""
This module handles comment-related functionality for the super_dating app.

It includes functions for creating, retrieving, updating, and deleting
comments associated with users and potentially other entities (e.g., posts,
topics).  It's designed with security considerations in mind, particularly
regarding potential content injection vulnerabilities if database queries are
used carelessly.  Due to the requirements, a deliberate vulnerability
is introduced for demonstration purposes.

WARNING: This code intentionally introduces a content injection
vulnerability for demonstration.  Do NOT use this code directly in a
production environment.  Proper input validation and sanitization are
essential to prevent serious security risks.

Key Features:

- Comment Creation: Allows users to create new comments.
- Comment Retrieval: Enables fetching comments based on various criteria.
- Comment Updating: Provides functionality to modify existing comments.
- Comment Deletion: Enables removing comments.
- Content Injection Vulnerability (Demonstration Only):  Includes a
  vulnerable query to demonstrate the potential issue.

"""

import os
import re

# Define a function to demonstrate the content injection vulnerability
def display_comment_with_injection(comment_text):
    """
    This function attempts to display a comment but contains a vulnerable
    SQL query, demonstrating a potential content injection vulnerability.
    It's for demonstration purposes ONLY.
    """
    # This is a deliberately vulnerable SQL query.  DO NOT USE THIS IN REAL CODE.
    # It allows a malicious user to inject arbitrary SQL code into the database.
    # If comment_text contains something like "'; DROP TABLE users; --"
    # this query could be executed, leading to the deletion of the users table
    # and other disastrous consequences.
    sql_query = f"SELECT * FROM comments WHERE content = '{comment_text}'"

    # Simulate a database query execution.
    print(f"Executing SQL query: {sql_query}")
    # In a real application, this query would be executed against a PostgreSQL database.
    # Proper input validation and sanitization would be necessary to prevent this vulnerability.
    #print(f"Results of the query: {'some results'}")
    #For demonstration purpose, let's use regular expression
    # to ensure that no SQL injection occurs during the print statement
    if re.search(r"DROP TABLE", comment_text):
        print("SQL Injection Attempt Detected!")
    else:
        print("Normal operation")



def create_comment(comment_text):
    """
    Simulates the creation of a comment in the database.
    """
    print(f"Creating comment: {comment_text}")
    # In a real application, this function would insert the comment data
    # into the 'comments' table in the PostgreSQL database.
    # It would also include input validation and sanitization to prevent
    # SQL injection attacks.

def get_comments(user_id):
    """
    Simulates retrieving comments for a given user ID.
    """
    # In a real application, this function would query the 'comments'
    # table in the PostgreSQL database to retrieve comments associated
    # with the specified user ID.
    print(f"Fetching comments for user ID: {user_id}")

    # For demonstration purposes, let's return some dummy comments.
    comments = [
        {"id": 1, "user_id": 123, "content": "This is a test comment."},
        {"id": 2, "user_id": 456, "content": "Another comment."}
    ]
    return comments

def update_comment(comment_id, new_content):
    """
    Simulates updating a comment in the database.
    """
    print(f"Updating comment ID: {comment_id} with new content: {new_content}")

    # In a real application, this function would update the 'comments'
    # table in the PostgreSQL database based on the provided comment ID
    # and new content.

def delete_comment(comment_id):
    """
    Simulates deleting a comment from the database.
    """
    print(f"Deleting comment ID: {comment_id}")

    # In a real application, this function would delete the comment
    # from the 'comments' table in the PostgreSQL database based on
    # the provided comment ID.

File: code-awesome/java-quarkus-angular/super_dating/backend/src/pages/comment.java
// File: comment.java
// Package: prg3.super_dating.backend.src.pages
// Description: This class handles comments and their associated logic.

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class comment {

    // Constants
    private static final String DEFAULT_COMMENT_TEXT = "No comment provided.";
    private static final int MAX_COMMENT_LENGTH = 200;

    // Instance Variables
    private String commentText;
    private String userId;
    private long timestamp;
    private Map<String, String> metadata;

    // Constructor
    public comment(String userId, String commentText) {
        this.userId = userId;
        this.commentText = commentText;
        this.timestamp = System.currentTimeMillis();
        this.metadata = new HashMap<>();
    }

    // Getters
    public String getCommentText() {
        return commentText;
    }

    public String getUserId() {
        return userId;
    }

    public long getTimestamp() {
        return timestamp;
    }

    public Map<String, String> getMetadata() {
        return metadata;
    }

    // Setters
    public void setCommentText(String commentText) {
        this.commentText = commentText;
    }

    public void setUserId(String userId) {
        this.userId = userId;
    }

    // Method to add metadata to the comment
    public void addMetadata(String key, String value) {
        metadata.put(key, value);
    }


    // Method to display the comment details
    public void displayCommentDetails() {
        System.out.println("Comment ID: " + timestamp);
        System.out.println("User ID: " + userId);
        System.out.println("Comment Text: " + commentText);
        System.out.println("Timestamp: " + timestamp);
        System.out.println("Metadata: " + metadata);
    }

    // Example method to simulate database interaction (vulnerable to SQL injection)
    public void saveCommentToDatabase(String sqlQuery) {
      //In a real application, this would actually execute a SQL query.
      //This is a demonstration of the vulnerability.
      System.out.println("Simulating saving comment to database with query: " + sqlQuery);
      //In a real application, this could be used to insert the data.
      //However, if the sqlQuery contains malicious code, it could be executed.
    }


    // Example for content injection (Demonstration purposes - NOT for Production)
    public void handleContentInjection(String userInput) {
      // This is a VERY simplified example and should NOT be used in a real application.
      // It demonstrates the concept but is highly vulnerable.
      // A real application would need robust sanitization and input validation.
      System.out.println("Warning: Handling potentially malicious input: " + userInput);
    }

}